fw/work_queue.c

```diff
@@ -46,12 +39,14 @@ tfw_wq_init(TfwRBQueue *q, int node)
 		atomic64_t *local_head = per_cpu_ptr(q->heads, cpu);
 		atomic64_set(local_head, LLONG_MAX);
 	}
+	q->qsize = qsize;
 	q->last_head = 0;
 	atomic64_set(&q->head, 0);
 	atomic64_set(&q->tail, 0);
 	set_bit(TFW_QUEUE_IPI, &q->flags);
 
-	q->array = kmalloc_node(QSZ * WQ_ITEM_SZ, GFP_KERNEL, node);
+	/* Fallback to vmalloc for large queue sizes, which might affect the performance */
+	q->array = kvmalloc_node(qsize * WQ_ITEM_SZ, GFP_KERNEL, node);
```

Comment:
Since v5.7 kernel, there shouldn't be page faults in runtime in vmalloc'ed area. See [this patch](https://lore.kernel.org/lkml/20200508144043.13893-1-joro@8bytes.org/).

Reply:
OK, then please update the confusing comment - it seems there should not be any performance impacts